gui/webdisplay/src/TWebWindow.cxx

```diff
@@ -220,6 +365,109 @@ void ROOT::Experimental::TWebWindow::InvokeCallbacks(bool force)
    }
 }
 
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Add key - procid pair for started window
+/// Key is random number generated when starting new window
+/// procid is special information about starting process which can be used later to halt it
+
+unsigned ROOT::Experimental::TWebWindow::AddProcId(bool batch_mode, const std::string &key, const std::string &procid)
+{
+   std::lock_guard<std::mutex> grd(fConnMutex);
+
+   ++fConnCnt;
+
+   fKeys.emplace_back(std::make_shared<WebConn>(fConnCnt, batch_mode, key, procid));
+
+   return fConnCnt;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Returns true if provided key value already exists (in processes map or in existing connections)
+
+bool ROOT::Experimental::TWebWindow::HasKey(const std::string &key)
+{
+   std::lock_guard<std::mutex> grd(fConnMutex);
+
+   for (auto &&entry : fKeys) {
+      if (entry->fKey == key)
+         return true;
+   }
+
+   for (auto &&conn : fConn) {
+      if (conn->fKey == key)
+         return true;
+   }
+
+   return false;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Check if started process(es) establish connection. After timeout such processed will be killed
+/// Method invoked from http server thread, therefore appropriate mutex must be used on all relevant data
+
+void ROOT::Experimental::TWebWindow::CheckWebKeys()
+{
+   if (!fMgr) return;
+
+   auto stamp = std::chrono::system_clock::now();
+
+   float tmout = fMgr->GetLaunchTmout();
+
+   std::vector<std::string> procs;
+
+   {
+      std::lock_guard<std::mutex> grd(fConnMutex);
+
+      for (auto n = fKeys.size(); n > 0; --n) {
+         std::chrono::duration<double> diff = stamp - fKeys[n - 1]->fSendStamp;
+         // introduce large timeout
+         if (diff.count() > tmout) {
+            R__DEBUG_HERE("webgui") << "Halt process " <<  fKeys[n - 1]->fProcId << " after " << diff.count() << " sec";
+            procs.emplace_back(fKeys[n - 1]->fProcId);
+            fKeys.erase(fKeys.begin() + n - 1);
+         }
+      }
+   }
+
+   for (auto &&entry : procs)
+      fMgr->HaltClient(entry);
+}
+
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Check if there are connection which are inactive for longer time
+/// For instance, batch browser will be stopped if no activity for 30 sec is there
+
+void ROOT::Experimental::TWebWindow::CheckInactiveConnections()
```

Comment:
See comment regarding `WaitForWebKeysAndHalt()`

Reply:
Same answer - is not waiting but cleanup inactive connection (if any)